### ISSUE TITLE
Add an API for sidecar.json overrides

### ISSUE THREAD

**Issue by a-hurst** [mne-tools/mne-bids#458]
#### Describe the problem
The BIDS-EEG specification has [a number of recommended & useful fields](https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/03-electroencephalography.html#sidecar-json-_eegjson) in the per-participant sidecar JSON files, including the institution where recording was made, the amplifier & cap manufacturer & model, the software used for data collection, any software or hardware filters applied to the data during recording, and more.

Most of these can't be inferred directly from the data files themselves, meaning they have to be manually specified at some point. For larger datasets, manually editing each participant's mne-bids generated sidecar to add this info ends up being slow and potentially error-prone.

#### Describe your solution
The easiest solution that comes to mind would be to add support for sidecar templates: users could create a .json file with all the sidecar fields they want to add or override, and then mne-bids would use that file to augment and/or correct the file produced by mne-bids. This approach is nice because it could easily be used with both the `raw_to_bids` CLI tool and the mne-bids Python module.

An example would look something like this:

```json
{
    "SamplingFrequency": 1000,
    "Manufacturer": "Compumedics Neuroscan",
    "ManufacturersModelName": "SynAmps 2/RT",
    "CapManufacturer": "Compumedics Neuroscan",
    "CapManufacturersModelName": "Quik-Cap 64ch",
    "SoftwareVersions": "Compumedics CURRY 7"
}
```

Note that "SamplingFrequency" is listed here even though it's auto-detected by MNE: that's because for my lab's amplifier (and presumably others) the sample rate is detected by MNE as 999.9999999999999 instead of 1000.0 and is printed as such.

#### Describe possible alternatives
An alternative idea would be for `write_raw_bids` to optionally return an intermediate BIDS object with the generated json/tsv info in a form that could be edited before being written out.

Basically, you'd break `write_raw_bids` internally into two functions: a `create_bids_obj` function and a `write_bids_obj` function, and create a public API for reading and editing that data pre-export. While more flexible, I'd imagine this approach would probably be more work and less pleasant to maintain.

**Comment by jasmainak** [comment]
I like the idea of sidecar templates! I think it's the neatest solution I've seen so far. So something like:

```py
update_sidecars(sidecar_template, 'regular_expression_of_jsons_to_overwrite')
```

?

can you write down a couple of lines of python code to expand on the API a bit more?

one issue I see is that there are sometimes fields in BIDS that are inter-linked. This could be potentially error prone that way. But we'll have to assume that users know what they are doing. Or if you have a better idea :-)

**Comment by a-hurst** [comment]
I was imagining something along the lines of an argument to the `write_raw_bids` function, e.g.

```python
template_path = "./setup_info.json"
write_raw_bids(
    raw_file, bids_basename, bids_root,
    events_data=events, sidecar_template=template_path
)
```
and via the CLI:
```bash
mne_bids raw_to_bids --subject_id sub01 --task rest --raw edfs/s01.edf --bids_root bids --sidecar_template setup_info.json
```

Then, when generating the OrderedDict containing the sidecar info to write out, `write_raw_bids` will read the key/value pairs from the sidecar template, add any missing values to the OrderedDict and overwrite any existing fields that have keys in the template (e.g. replacing the auto-generated`Manufacturer: "n/a"` with `Manufacturer: "BioSemi"` from the template), and then write out the resulting file.

This would probably also benefit from a list of valid sidecar fields to check the sidecar template against, to make sure all the fields are valid BIDS names that don't contain any typos.

**Comment by jasmainak** [comment]
> Note that "SamplingFrequency" is listed here even though it's auto-detected by MNE: that's because for my lab's amplifier (and presumably others) the sample rate is detected by MNE as 999.9999999999999 instead of 1000.0 and is printed as such.

By the way you can fix this already using the current API by doing:

```py
raw.info['sfreq'] = 1000.0
write_raw_bids(raw, bids_basename, ...)
```

in case you weren't aware

**Comment by a-hurst** [comment]
Thanks! Still learning the ropes in MNE so that's helpful.

**Comment by a-hurst** [comment]
Okay, just took a stab at actually implementing this. It's not documented or particularly well tested, but it seems to work! You can check out the changes to write.py in my sidecar_template branch [here](https://github.com/mne-tools/mne-bids/compare/master...a-hurst:sidecar_template).

Basically, you pass a path to a valid .json file to `write_raw_bids`' new sidecar_template argument. Then, when generating the sidecar, it will:

1. Generate the OrderedDict of sidecar.json fields, as per usual
2. Load in the template as an OrderedDict
3. Update pre-generated OrderedDict with keys from the template dict, provided their values are not null (the reason for this last bit is explained below)

I've also written this so it's possible to specify the order of the sidecar fields using the template. This is mostly so you can group pre-generated fields with related template fields (e.g. "Manufacturer" with "ManfacturersModelName") in the file to keep them organized and easily readable. Basically, if you add a field to the template that has a `null` value, and that field matches the name of a pre-generated field (e.g. "PowerLineFrequency"), the pre-generated field and its value will be reported in that location in the sidecar.

To illustrate, here's an example sidecar template, based on one of the BIDS-EEG example datasets I found:

```json
{
    "SamplingFrequency": null,
    "Manufacturer": "Brain Products",
    "ManufacturersModelName": "BrainAmp DC",
    "CapManufacturer": "Brain Products",
    "CapManufacturersModelName": "actiCAP 64Ch Standard-2",
    "SoftwareVersions": "Labstreaming Layer BrainAmpSeries 1.11",
    "EEGChannelCount": null,
    "EOGChannelCount": null,
    "ECGChannelCount": null,
    "EMGChannelCount": null,
    "MiscChannelCount": null,
    "TriggerChannelCount": null,
    "PowerLineFrequency": null,
    "EEGPlacementScheme": "10 percent system",
    "EEGReference": "unipolar, placed on Fz",
    "EEGGround": "placed on Afz",
    "SoftwareFilters": "n/a",
    "RecordingDuration": null,
    "RecordingType": null
}
```

When passed to `write_raw_bids`, the resulting sidecar will look something like this:

```json
{
    "TaskName": "matchingpennies",
    "SamplingFrequency": 5000,
    "Manufacturer": "Brain Products",
    "ManufacturersModelName": "BrainAmp DC",
    "CapManufacturer": "Brain Products",
    "CapManufacturersModelName": "actiCAP 64Ch Standard-2",
    "SoftwareVersions": "Labstreaming Layer BrainAmpSeries 1.11",
    "EEGChannelCount": 10,
    "EOGChannelCount": 0,
    "ECGChannelCount": 0,
    "EMGChannelCount": 0,
    "MiscChannelCount": 0,
    "TriggerChannelCount": 1,
    "PowerLineFrequency": 50,
    "EEGPlacementScheme": "10 percent system",
    "EEGReference": "unipolar, placed on Fz",
    "EEGGround": "placed on Afz",
    "SoftwareFilters": "n/a",
    "RecordingDuration": 2345.9518,
    "RecordingType": "continuous"
}
```

With the null values overridden by their generated ones, and any fields not specified in the template appearing at the top of the sidecar.

Let me know what you think!

**Comment by jasmainak** [comment]
@a-hurst thanks for taking a stab at this already! For the future, I would suggest discussing the API a bit before attempting an implementation. I don't want to make you do double effort :) However, since you already have some code, please make a PR so that you can get feedback early.

I'm a bit partial to a separate function. The advantage is that you do not need a `raw` object to update the sidecars so the can be used to make "fixes" in the dataset. With the current API, it will become unwieldy and slow if people have to update something and load the `raw` objects again. Let me know what you think.

As for ordering the values, I would say alphabetic should be good enough. Keep it simple stupid :) If you don't want to overwrite a value, just leave it out of the template.

Might be good to get some more opinions perhaps @sappelhoff @hoechenberger @adam2392 ?

**Comment by sappelhoff** [comment]
sounds really nice to me and like something that we need.

> I'm a bit partial to a separate function. The advantage is that you do not need a raw object to update the sidecars so the can be used to make "fixes" in the dataset. With the current API, it will become unwieldy and slow if people have to update something and load the raw objects again. Let me know what you think.

I agree with @jasmainak that something like

`update_sidecars(bids-path, JSON)`

would probably be more flexible. So via the CLI that'd be something like:

```python
mne_bids raw_to_bids --subject_id 01 --task REST --raw 01.edf --bids_root PATH 
mne_bids update_sidecars --bids_root PATH --sidecar_template MY_JSON
```

> As for ordering the values, I would say alphabetic should be good enough.

I agree that the ordering is just a "nice addon", and we should focus on the more important functionality.

> This would probably also benefit from a list of valid sidecar fields to check the sidecar template against, to make sure all the fields are valid BIDS names that don't contain any typos.

I think that this should be done by the validator. After every conversion, a user should run the bids-validator anyways ... and then they will be alerted that something is wrong. 

--> I dream about a Python bids-validator many times a day, but we should keep bids-validation in one place, I think. (and not replicate *some* validation within mne-bids ... despite the fact that it'd be easier for us to fix and extend, because it's not convoluted+undocumented JavaScript code :slightly_smiling_face: )

**Comment by a-hurst** [comment]
> @a-hurst thanks for taking a stab at this already! For the future, I would suggest discussing the API a bit before attempting an implementation. I don't want to make you do double effort :) However, since you already have some code, please make a PR so that you can get feedback early.

No worries, I sort of needed the functionality to move forward on a project anyway so it was time well-spent regardless! I thought a quick-and-dirty implementation would help people decide whether they liked the general API too, which it seems like it has.

You're right that a separate function makes the most sense (hadn't thought about updating existing datasets). The code would still be pretty much the same regardless, except instead of updating the OrderedDict before it's generated we'd be reading the sidecar JSON to an OrderedDict and updating that with the template. I can probably find time to implement this later today and make a proper PR.

> As for ordering the values, I would say alphabetic should be good enough. Keep it simple stupid :) If you don't want to overwrite a value, just leave it out of the template.

It might be simpler code-wise, but I think you might be underestimating the overlap between "I want my data to be meticulously organized into logical groups" people (i.e. most BIDS users) and "I want my metadata files to be meticulously organized into logical groups" people 😉.

Also although there's nothing in the BIDS standard I found about sequence of fields, all the example sidecars I've seen have sorts of unspoken conventions like "TaskName" being at the top and optionally followed by "TaskDescription" and "InstitutionName", "SoftwareFilters" & "HardwareFilters" being next to each other, etc. that blind alphabetical sorting would break. I can try to think of a cleaner alternative to my current approach, though.

**Comment by sappelhoff** [comment]
> Also although there's nothing in the BIDS standard I found about sequence of fields

indeed, there is nothing regarding the order of fields in JSON files.

We do have a bit about ordering or columns and rows in TSV files (e.g., channels tsv files). But some of these rules are currently not tested by the BIDS validator either (e.g., that in channels TSV files, where each row is a channel, the order should be as it is in the data file)

**Comment by ihgumilar** [comment]
Hi, 

I am wondering about the make_bids_basename. Is it possible if I want to add other parameters that are not covered by make_bids_basename (eg. place of recording), instead of only using parameters that are provided by make_bids_basename (eg. subject, session, task, suffix) ?

In other words, is it possible if I add metadata that is not defined by make_bids_basename ?

Is this thread similar to what I am asking ? sorry about this because I am new to BIDS :)

Thanks for your help

**Comment by jasmainak** [comment]
@ihgumilar for usage questions gitter might be better: https://gitter.im/mne-tools/mne-bids

re your question I am not sure what exactly you have in mind. Are you talking about MEG/EEG/iEEG part of the BIDS specification? `make_bids_basename` should cover all the parameters for these specs

**Comment by adam2392** [comment]
@a-hurst did you ever end up having some prelim working code for this? Would you want to start the PR for this? I would love to take a look to see how things are being done and help out to get it merged because I'm running into a similar issue.

**Comment by a-hurst** [comment]
@adam2392 Did you check out my `sidecar_template` branch ([found here](https://github.com/mne-tools/mne-bids/compare/master...a-hurst:sidecar_template))? It definitely needs some work, especially since the maintainers had some valid issues with my approach (see the above conversation), but it was complete enough to get my dataset converted to BIDS with more-or-less all the details I wanted organized the way I wanted. I think I have a local branch with a bit more recent work, but nothing major.

If you've got the time and will to turn this into a proper PR that would be fantastic: I'd do it myself, but I just finished a cross-country move and am also trying to finish up my thesis, so my spare time is at a minimum right now!
